File: src/structural/bridge3.ts
abstract class Pizza {
  protected crust: string = "Thin";
  protected sauce: string = "Tomato sauce";
  protected toppings: string | null = null;

  abstract setCrust(crust: string): void;
  abstract setSauce(sauce: string): void;
  abstract setToppings(toppings: string | null): void;

  abstract assemble(): void;
  abstract qualityCheck(): void;
}

class PepperoniPizza extends Pizza {
  setToppings(toppings: string | null): void {
    this.toppings = toppings;
  }

  setSauce(sauce: string): void {
    this.sauce = sauce;
  }

  setCrust(crust: string): void {
    this.crust = crust;
  }

  assemble(): void {
    console.log("Adding Sauce: " + this.sauce);
    console.log(this.toppings ? "Adding Toppings: " + this.toppings : "No toppings added");
    console.log("Adding Pepperoni");
  }

  qualityCheck(): void {
    console.log("Crust is: " + this.crust);
  }
}

class VeggiePizza extends Pizza {
  setToppings(toppings: string | null): void {
    this.toppings = toppings;
  }

  setSauce(sauce: string): void {
    this.sauce = sauce;
  }

  setCrust(crust: string): void {
    this.crust = crust;
  }

  assemble(): void {
    console.log("Adding Sauce: " + this.sauce);
    console.log(this.toppings ? "Adding Toppings: " + this.toppings : "No toppings added");
    console.log("Adding Cheese");
  }

  qualityCheck(): void {
    console.log("Crust is: " + this.crust);
  }
}

abstract class Restaurant {
  constructor(protected pizza: Pizza) {}

  abstract makeCrust(): void;
  abstract addSauce(): void;
  abstract addToppings(): void;

  deliver(): void {
    this.makeCrust();
    this.addSauce();
    this.addToppings();
    this.pizza.assemble();
    this.pizza.qualityCheck();
    console.log("Order in Progress!");
  }
}

class ItalianRestaurant extends Restaurant {
  constructor(pizza: Pizza) {
    super(pizza);
  }

  makeCrust(): void {
    this.pizza.setCrust("Thin");
  }

  addSauce(): void {
    this.pizza.setSauce("Oil");
  }

  addToppings(): void {
    this.pizza.setToppings(null);
  }
}

class AmericanRestaurant extends Restaurant {
  constructor(pizza: Pizza) {
    super(pizza);
  }

  makeCrust(): void {
    this.pizza.setCrust("Thick");
  }

  addSauce(): void {
    this.pizza.setSauce("Super Secret Recipe");
  }

  addToppings(): void {
    this.pizza.setToppings("Everything");
  }
}

function makePizza(): void {
  const americanRestaurant = new AmericanRestaurant(new PepperoniPizza());
  americanRestaurant.deliver();

  console.log("---");

  const italianRestaurant = new ItalianRestaurant(new VeggiePizza());
  italianRestaurant.deliver();
}

makePizza();

File: src/structural/bridge3.java
@Data
public abstract class Pizza {
    protected String sauce;
    protected String toppings;
    protected String crust;

    public abstract void assemble();
    public abstract void qualityCheck();
}

public class PepperoniPizza extends Pizza {
    @Override
    public void assemble() {
        System.out.println("Adding Sauce: " + sauce);
        System.out.println("Adding Toppings: " + toppings);
        System.out.println("Adding Pepperoni");
    }

    @Override
    public void qualityCheck() {
        System.out.println("Crust is: " + crust);
    }
}

public class VeggiePizza extends Pizza {
    @Override
    public void assemble() {
        System.out.println("Adding Sauce: " + sauce);
        System.out.println("Adding Toppings: " + toppings);
        System.out.println("Adding Cheese");
    }

    @Override
    public void qualityCheck() {
        System.out.println("Crust is: " + crust);
    }
}

public abstract class Restaurant {
    protected Pizza pizza;

    protected Restaurant(Pizza pizza) {
        this.pizza = pizza;
    }

    abstract void addSauce();
    abstract void addToppings();
    abstract void makeCrust();

    public void deliver() {
        makeCrust();
        addSauce();
        addToppings();
        pizza.assemble();
        pizza.qualityCheck();
        System.out.println("Order in Progress!");
    }
}

public class ItalianRestaurant extends Restaurant {
    public ItalianRestaurant(Pizza pizza) {
        super(pizza);
    }

    @Override
    public void addToppings() {
        pizza.setToppings(null);
    }

    @Override
    public void addSauce() {
        pizza.setSauce("Oil");
    }

    @Override
    public void makeCrust() {
        pizza.setCrust("Thin");
    }
}

public class AmericanRestaurant extends Restaurant {
    public AmericanRestaurant(Pizza pizza) {
        super(pizza);
    }

    @Override
    public void addToppings() {
        pizza.setToppings("Everything");
    }

    @Override
    public void addSauce() {
        pizza.setSauce("Super Secret Recipe");
    }

    @Override
    public void makeCrust() {
        pizza.setCrust("Thick");
    }
}

public static void main(String[] args) {
    AmericanRestaurant americanRestaurant = new AmericanRestaurant(new PepperoniPizza());
    americanRestaurant.deliver();

    ItalianRestaurant italianRestaurant = new ItalianRestaurant(new VeggiePizza());
    italianRestaurant.deliver();
}